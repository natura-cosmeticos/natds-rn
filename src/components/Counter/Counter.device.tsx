import { storiesOf } from '@storybook/react-native'
import {
  All,
  Default,
  Disabled,
  Interactive,
  Label,
  Size,
  Value
} from './Counter.stories'

storiesOf('Counter', module)
  .add('All', All)
  .add('Default', Default)
  .add('Label', Label)
  .add('Size', Size)
  .add('Disabled', Disabled)
  .add('Value', Value)
  .add('Interactive', Interactive)
