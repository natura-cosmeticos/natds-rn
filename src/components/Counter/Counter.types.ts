import { Size } from '@naturacosmeticos/natds-themes/react-native';

export type CounterSizes = keyof Pick<Size, 'medium' | 'semiX'>;

export interface CounterProps {
  /**
   * An accessibility hint helps users understand what will happen when they perform an action
   * on the accessibility element when that result is not clear from the accessibility label.
   */
  decrementButtonAccessibilityHint?: string,
  /**
   * An accessibility label overrides the text that's read by the screen reader when the user
   * interacts with the element. By default, the label is constructed by traversing all the
   * children and accumulating all the Text nodes separated by space.
   */
  decrementButtonAccessibilityLabel?: string,
  /**
   * An accessibility hint helps users understand what will happen when they perform an action
   * on the accessibility element when that result is not clear from the accessibility label.
   */
  incrementButtonAccessibilityHint?: string,
  /**
   * An accessibility label overrides the text that's read by the screen reader when the user
   * interacts with the element. By default, the label is constructed by traversing all the
   * children and accumulating all the Text nodes separated by space.
   */
  incrementButtonAccessibilityLabel?: string,
  /**
   * An accessibility hint helps users understand what will happen when they perform an action
   * on the accessibility element when that result is not clear from the accessibility label.
   */
  inputAccessibilityHint?: string,
  /**
   * An accessibility label overrides the text that's read by the screen reader when the user
   * interacts with the element. By default, the label is constructed by traversing all the
   * children and accumulating all the Text nodes separated by space.
   */
  inputAccessibilityLabel?: string,
  /**
   * When `true` both the decrement and increment button are disabled.
   * @default false
   */
  disabled?: boolean,
  /**
   * When provided this shows above the counter.
   */
  label?: string,
  /**
   * The minimum valid value for the component
   * @default 0
   */
  minValue?: number,
  /**
   * The maximum valid value for the component
   * @default 99
   */
  maxValue?: number,
  /**
   * Callback that is called when the decrement button is pressed.
   */
  onDecrement: () => void,
  /**
   * Callback that is called when the increment button is pressed.
   */
  onIncrement: () => void,
  /**
   * This defines the height of the component, according to
   * corresponding theme Size token.
   * @default 'medium'
   */
  size?: CounterSizes,
  /**
   * Optional ID for testing.
   */
  testID?: string,
  /**
   * Current value of the counter.
   */
  value?: number,
}
