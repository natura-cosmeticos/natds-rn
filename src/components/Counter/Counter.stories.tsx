/* eslint-disable max-lines */
import React from 'react'
import { boolean, number, text } from '@storybook/addon-knobs'
import { Text } from 'react-native'
import { StoryContainer, VerticalStoryContainer } from '../../common/HelperComponents/StoryContainer'
import { Counter } from './Counter'

const description = () => `
---

  ### NOTE:
  This component is available in the following variants:

  - ✅ Standard

With the following attribute status:

  - ✅ **Label**
  - **Size:**
    - ✅ \`Medium\`
    - ✅ \`SemiX\`
  - ✅ **Disabled**

---
`

export default {
  component: Counter,
  parameters: {
    componentSubtitle: 'A Counter reduces input effort for fields with values that deviate little from the default by allowing users to increase or decrease the number in a single button press.',
    docs: {
      extractComponentDescription: description
    }
  },
  title: 'Components|Counter'
}

export const Default = () => {
  const [value, setValue] = React.useState(0)

  return (
    <StoryContainer title="Default">
      <Counter
        onDecrement={() => setValue(value - 1)}
        onIncrement={() => setValue(value + 1)}
        value={value}
      />
    </StoryContainer>
  )
}

export const Label = () => {
  const [value, setValue] = React.useState(0)

  return (
    <StoryContainer title="With label">
      <Counter
        decrementButtonAccessibilityHint="decrement the value of the counter"
        decrementButtonAccessibilityLabel="decrement button"
        incrementButtonAccessibilityHint="increment the value of the counter"
        incrementButtonAccessibilityLabel="increment button"
        inputAccessibilityHint="you can set the quantity you want editing this field"
        inputAccessibilityLabel={`you have ${value} items in your shopping cart`}
        onDecrement={() => setValue(value - 1)}
        onIncrement={() => setValue(value + 1)}
        label="Label"
        value={value}
      />
    </StoryContainer>
  )
}

export const Size = () => {
  const [mediumValue, setMediumValue] = React.useState(0)
  const [semiXValue, setSemiXValue] = React.useState(0)

  return (
    <StoryContainer title="Sizes">
      <Counter
        decrementButtonAccessibilityHint="decrement the value of the counter"
        decrementButtonAccessibilityLabel="decrement button"
        incrementButtonAccessibilityHint="increment the value of the counter"
        incrementButtonAccessibilityLabel="increment button"
        inputAccessibilityHint="you can set the quantity you want editing this field"
        inputAccessibilityLabel={`you have ${mediumValue} items in your shopping cart`}
        onDecrement={() => setMediumValue(mediumValue - 1)}
        onIncrement={() => setMediumValue(mediumValue + 1)}
        label="medium"
        size="medium"
        value={mediumValue}
      />
      <Counter
        decrementButtonAccessibilityHint="decrement the value of the counter"
        decrementButtonAccessibilityLabel="decrement button"
        incrementButtonAccessibilityHint="increment the value of the counter"
        incrementButtonAccessibilityLabel="increment button"
        inputAccessibilityHint="you can set the quantity you want editing this field"
        inputAccessibilityLabel={`you have ${semiXValue} items in your shopping cart`}
        onDecrement={() => setSemiXValue(semiXValue - 1)}
        onIncrement={() => setSemiXValue(semiXValue + 1)}
        label="semiX"
        size="semiX"
        value={semiXValue}
      />
    </StoryContainer>
  )
}
export const Disabled = () => {
  const [value, setValue] = React.useState(0)

  return (
    <StoryContainer title="Disabled">
      <Counter
        decrementButtonAccessibilityHint="decrement the value of the counter"
        decrementButtonAccessibilityLabel="decrement button"
        incrementButtonAccessibilityHint="increment the value of the counter"
        incrementButtonAccessibilityLabel="increment button"
        inputAccessibilityHint="you can set the quantity you want editing this field"
        inputAccessibilityLabel={`you have ${value} items in your shopping cart`}
        onDecrement={() => setValue(value - 1)}
        onIncrement={() => setValue(value + 1)}
        value={value}
        disabled
      />
    </StoryContainer>
  )
}

export const Value = () => {
  const initialValue = 42
  const [value, setValue] = React.useState(initialValue)

  return (
    <VerticalStoryContainer title="Value">
      <Text style={{ paddingBottom: 16 }}>
        You can set a minimum, maximum and initial value for the component
      </Text>
      <Counter
        decrementButtonAccessibilityHint="decrement the value of the counter"
        decrementButtonAccessibilityLabel="decrement button"
        incrementButtonAccessibilityHint="increment the value of the counter"
        incrementButtonAccessibilityLabel="increment button"
        inputAccessibilityHint="this shows the quantity of items you have in your shopping cart"
        inputAccessibilityLabel={`you have ${value} items in your shopping cart`}
        label="Fixed range of value"
        onDecrement={() => setValue(value - 1)}
        onIncrement={() => setValue(value + 1)}
        maxValue={45}
        minValue={40}
        value={value}
      />
    </VerticalStoryContainer>
  )
}

export const Interactive = () => {
  const [value, setValue] = React.useState(0)

  return (
    <StoryContainer title="Interactive">
      <Counter
        decrementButtonAccessibilityHint={text('Decrement button Accessibility Hint', 'A hint about what the decrement button does')}
        decrementButtonAccessibilityLabel={text('Decrement button Accessibility Label', 'Description of what the decrement button does')}
        incrementButtonAccessibilityHint={text('Increment button Accessibility Hint', 'A hint about what the increment button does')}
        incrementButtonAccessibilityLabel={text('Increment button Accessibility Label', 'Description of what the increment button does')}
        inputAccessibilityHint={text('Input Accessibility Hint', 'A hint about what the input does')}
        inputAccessibilityLabel={text('Input Accessibility Label', 'Description of what the input does')}
        disabled={boolean('Disabled', false)}
        label={text('Label', 'Interactive example')}
        maxValue={number('Maximum value for the counter', 99)}
        minValue={number('Minimum value for the counter', 0)}
        onDecrement={() => setValue(value - 1)}
        onIncrement={() => setValue(value + 1)}
        value={value}
      />
    </StoryContainer>
  )
}

export const All = () => (
  <>
    <Default />
    <Label />
    <Size />
    <Disabled />
    <Value />
  </>
)
