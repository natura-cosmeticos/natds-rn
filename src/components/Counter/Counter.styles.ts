/* eslint-disable max-len */
import { Theme } from '@naturacosmeticos/natds-themes/react-native'
import { CSSObject } from 'styled-components'
import styled from 'styled-components/native'
import {
  getColorHighEmphasis,
  getColorLowEmphasis,
  getColorMediumEmphasis,
  getColorSurface,
  getSize,
  getSizeMediumX,
  getSpacingMicro,
  getSpacingSmall
} from '../../common/themeSelectors'
import { CounterSizes } from './Counter.types'

type CounterStyleProps = {
  theme: Theme;
}

type ContainerStyleProps = CounterStyleProps & {
  size: CounterSizes;
}

export const Label = styled.Text<{ theme: Theme }>(({ theme }): CSSObject => ({
  color: getColorMediumEmphasis(theme),
  fontFamily: theme.counter.label.primary.fontFamily,
  fontSize: theme.counter.label.fontSize,
  fontWeight: theme.counter.label.primary.fontWeight,
  letterSpacing: theme.counter.label.letterSpacing,
  lineHeight: theme.counter.label.fontSize * theme.counter.label.lineHeight,
  marginBottom: getSpacingMicro(theme)
}))

export const Container = styled.View<ContainerStyleProps>(({ size, theme }: ContainerStyleProps): CSSObject => ({
  alignItems: 'center',
  alignSelf: 'flex-start',
  backgroundColor: getColorSurface(theme),
  borderColor: getColorLowEmphasis(theme),
  borderRadius: theme.counter.borderRadius,
  borderWidth: 1,
  flexDirection: 'row',
  height: getSize(theme, size)
}))

export const Input = styled.TextInput<{ theme: Theme }>(({ theme }): CSSObject => ({
  borderColor: getColorLowEmphasis(theme),
  borderLeftWidth: 1,
  borderRightWidth: 1,
  color: getColorHighEmphasis(theme),
  fontFamily: theme.counter.content.primary.fontFamily,
  fontSize: theme.counter.content.fontSize,
  fontWeight: theme.counter.content.primary.fontWeight,
  height: '100%',
  letterSpacing: theme.counter.content.letterSpacing,
  lineHeight: theme.counter.content.fontSize * theme.counter.content.lineHeight,
  paddingHorizontal: getSpacingSmall(theme),
  textAlign: 'center',
  width: getSizeMediumX(theme)
}))
