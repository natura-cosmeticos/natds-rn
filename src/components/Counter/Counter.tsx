import React from 'react'
import { View } from 'react-native'
import { Button } from '../Button'
import { Container, Input, Label } from './Counter.styles'
import { CounterProps } from './Counter.types'

export const Counter = ({
  decrementButtonAccessibilityHint,
  decrementButtonAccessibilityLabel,
  incrementButtonAccessibilityHint,
  incrementButtonAccessibilityLabel,
  inputAccessibilityHint,
  inputAccessibilityLabel,
  disabled,
  label,
  minValue = 0,
  maxValue = 99,
  onDecrement,
  onIncrement,
  size = 'medium',
  testID = 'counter',
  value = 0
}: CounterProps) => (
  <View>
    { label
        && (
        <Label testID="counter-label">
          { label }
        </Label>
        )}
    <Container size={size} testID={testID} accessibilityLiveRegion="assertive">
      <Button
        accessibilityHint={decrementButtonAccessibilityHint}
        accessibilityLabel={decrementButtonAccessibilityLabel}
        size={size}
        type="text"
        text="âˆ’"
        onPress={onDecrement}
        disabled={disabled || value <= minValue}
      />
      <Input
        accessibilityHint={inputAccessibilityHint}
        accessibilityLabel={inputAccessibilityLabel}
        editable={false}
        value={value.toString()}
        testID="counter-input"
      />
      <Button
        accessibilityHint={incrementButtonAccessibilityHint}
        accessibilityLabel={incrementButtonAccessibilityLabel}
        size={size}
        type="text"
        text="+"
        onPress={onIncrement}
        disabled={disabled || value >= maxValue}
      />
    </Container>
  </View>
)
