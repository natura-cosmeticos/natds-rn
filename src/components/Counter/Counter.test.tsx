import { fireEvent } from '@testing-library/react-native';
import React from 'react';
import { renderWithTheme } from '../../../test/testHelpers';
import { Counter } from './Counter';
import mockTheme from '../../common/themeSelectors/theme/mock-theme.json';

const defaultProps = {
  onDecrement: () => {},
  onIncrement: () => {},
};

describe('Counter', () => {
  it('should render with default props', () => {
    const { toJSON } = renderWithTheme(<Counter {...defaultProps} />);

    expect(toJSON()).toMatchSnapshot();
  });
  describe('label attribute', () => {
    it('should render without label by default', () => {
      const { queryByTestId } = renderWithTheme(<Counter {...defaultProps}/>);

      expect(queryByTestId('counter-label')).toBeFalsy();
    });
    it('should render with given label', () => {
      const { getByTestId } = renderWithTheme(<Counter {...defaultProps} label="testing label" />);

      expect(getByTestId('counter-label')).toHaveTextContent('testing label');
    });
  });
  describe('size attribute', () => {
    it('should render with default size', () => {
      const { getByTestId } = renderWithTheme(<Counter {...defaultProps} />);

      expect(getByTestId('counter')).toHaveStyle({ height: mockTheme.size.medium });
    });
    it('should render with given size', () => {
      const { getByTestId } = renderWithTheme(<Counter {...defaultProps} size="semiX" />);

      expect(getByTestId('counter')).toHaveStyle({ height: mockTheme.size.semiX });
    });
  });
  describe('disabled attribute', () => {
    it('should render with decrement button disabled and increment button enabled by default', () => {
      const { getByText } = renderWithTheme(<Counter {...defaultProps} />);

      expect(getByText('−')).toBeDisabled();
      expect(getByText('+')).toBeEnabled();
    });
    it('should render with enabled buttons when value is between minimum and maximum values', () => {
      const { getByText } = renderWithTheme(
        <Counter {...defaultProps} minValue={1} value={2} maxValue={3} />,
      );

      expect(getByText('−')).toBeEnabled();
      expect(getByText('+')).toBeEnabled();
    });
    it('should disable decrement button when value is equal to minimum', () => {
      const { getByText } = renderWithTheme(
        <Counter {...defaultProps} minValue={1} value={1} maxValue={3} />,
      );

      expect(getByText('−')).toBeDisabled();
      expect(getByText('+')).toBeEnabled();
    });
    it('should disable increment button when value is equal to maximum', () => {
      const { getByText } = renderWithTheme(
        <Counter {...defaultProps} minValue={1} value={3} maxValue={3} />,
      );

      expect(getByText('−')).toBeEnabled();
      expect(getByText('+')).toBeDisabled();
    });
    it('should render with disabled buttons if disabled', () => {
      const { getByText } = renderWithTheme(<Counter {...defaultProps} disabled />);

      expect(getByText('−')).toBeDisabled();
      expect(getByText('+')).toBeDisabled();
    });
  });
  describe('value handling', () => {
    it('should render with value 0 as default', () => {
      const { getByTestId } = renderWithTheme(<Counter {...defaultProps} />);

      expect(getByTestId('counter-input').props.value).toBe('0');
    });
    it('should render with given value', () => {
      const { getByTestId } = renderWithTheme(<Counter {...defaultProps} value={10} />);

      expect(getByTestId('counter-input').props.value).toBe('10');
    });
  });
  describe('actions', () => {
    it('should call `onDecrement` when decrement button is pressed and value is greater than minimum value', () => {
      const onDecrement = jest.fn();
      const { getByText } = renderWithTheme(
        <Counter
          onIncrement={defaultProps.onIncrement}
          onDecrement={onDecrement}
          value={1}
        />,
      );

      fireEvent.press(getByText('−'));

      expect(onDecrement).toHaveBeenCalled();
    });
    it('should call `onIncrement` when increment button is pressed and value is less than maximum value', () => {
      const onIncrement = jest.fn();
      const { getByText } = renderWithTheme(
        <Counter
          onIncrement={onIncrement}
          onDecrement={defaultProps.onDecrement}
          value={1}
        />,
      );

      fireEvent.press(getByText('+'));

      expect(onIncrement).toHaveBeenCalled();
    });
  });
});
