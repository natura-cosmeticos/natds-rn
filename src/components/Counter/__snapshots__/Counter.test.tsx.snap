// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should render with default props 1`] = `
<View>
  <View
    accessibilityLiveRegion="assertive"
    size="medium"
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "backgroundColor": "#ffffff",
          "borderColor": "#bbbbbb",
          "borderRadius": 4,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 48,
        },
      ]
    }
    testID="counter"
  >
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderRadius": 4,
        }
      }
      testID="touchable-ripple"
    >
      <View
        accessibilityRole="button"
        disabled={true}
        size="medium"
        style={
          Array [
            Object {
              "alignContent": "center",
              "alignItems": "center",
              "backgroundColor": "#FFFFFF00",
              "borderColor": "#FFFFFF00",
              "borderRadius": 4,
              "borderWidth": 0,
              "justifyContent": "center",
              "minHeight": 48,
              "paddingHorizontal": 16,
            },
            Object {},
          ]
        }
        testID="button"
        type="text"
      >
        <View
          iconPosition="right"
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              },
            ]
          }
        >
          <Text
            disabled={true}
            iconPosition="right"
            style={
              Array [
                Object {
                  "color": "#777777",
                  "fontFamily": "Roboto",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "letterSpacing": 1.23,
                  "lineHeight": 19,
                  "marginEnd": 0,
                  "marginStart": 0,
                },
              ]
            }
            testID="button-label"
            type="text"
          >
            âˆ’
          </Text>
        </View>
      </View>
    </View>
    <TextInput
      editable={false}
      style={
        Array [
          Object {
            "borderColor": "#bbbbbb",
            "borderLeftWidth": 1,
            "borderRightWidth": 1,
            "color": "#333333",
            "fontFamily": "Roboto",
            "fontSize": 16,
            "fontWeight": "400",
            "height": "100%",
            "letterSpacing": 0.51,
            "lineHeight": 20,
            "paddingHorizontal": 16,
            "textAlign": "center",
            "width": 56,
          },
        ]
      }
      testID="counter-input"
      value="0"
    />
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderRadius": 4,
        }
      }
      testID="touchable-ripple"
    >
      <View
        accessibilityRole="button"
        disabled={false}
        size="medium"
        style={
          Array [
            Object {
              "alignContent": "center",
              "alignItems": "center",
              "backgroundColor": "#FFFFFF00",
              "borderColor": "#FFFFFF00",
              "borderRadius": 4,
              "borderWidth": 0,
              "justifyContent": "center",
              "minHeight": 48,
              "paddingHorizontal": 16,
            },
            Object {},
          ]
        }
        testID="button"
        type="text"
      >
        <View
          iconPosition="right"
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              },
            ]
          }
        >
          <Text
            disabled={false}
            iconPosition="right"
            style={
              Array [
                Object {
                  "color": "#333333",
                  "fontFamily": "Roboto",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "letterSpacing": 1.23,
                  "lineHeight": 19,
                  "marginEnd": 0,
                  "marginStart": 0,
                },
              ]
            }
            testID="button-label"
            type="text"
          >
            +
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
